### PR TITLE
Upated GCC to 9.3.0

### PR DESCRIPTION
**Pull request by dbeef**
I created patch for GCC 9.3.0, based on [top-sekret's 8.3.0 patch](https://github.com/top-sekret/psptoolchain/blob/master/patches/gcc-8.3.0-PSP.patch), which in turn is based on the original 4.9.3 patch.

Everything seems to work.
I had to do two troubling changes in order to make it build:

1:
https://github.com/dbeef/psptoolchain/commit/143e0e986056e75ba0be2a96b83a47877925c41e#r38540815
2:
https://github.com/dbeef/psptoolchain/commit/143e0e986056e75ba0be2a96b83a47877925c41e#r38540809

This is considered as work-in-progress until more tests are done.

### CODE REVIEW

**Comment by dbeef** [comment]
There are hints suggesting that CPP standard library has not been built correctly.
I attach Spelunky-PSP build log on GCC 9.3.0:

```
[ 11%] Building C object vendor/glad/glad_opengl_es_1.0/CMakeFiles/glad.dir/src/glad.c.obj
cd /home/dbeef/spelunky-psp/tmp/build-psp/vendor/glad/glad_opengl_es_1.0 && /usr/local/pspdev/bin/psp-gcc -DHAVE_OPENGL -D__PSP__ -Dwchar_t=short -I/home/dbeef/spelunky-psp/.. -I/usr/local/pspdev/psp/sdk/include -I/home/dbeef/spelunky-psp/vendor/glad/glad_opengl_es_1.0/include  -D_PSP_FW_VERSION=999999 -O3 -DNDEBUG   -G0 -I/usr/local/pspdev/psp/include -I/usr/local/pspdev/psp/sdk/include/libc -G0 -I/usr/local/pspdev/psp/include -I/usr/local/pspdev/psp/sdk/include/libc -std=gnu11 -o CMakeFiles/glad.dir/src/glad.c.obj   -c /home/dbeef/spelunky-psp/vendor/glad/glad_opengl_es_1.0/src/glad.c
/home/dbeef/spelunky-psp/vendor/glad/glad_opengl_es_1.0/src/glad.c: In function ‘free_exts’:
/home/dbeef/spelunky-psp/vendor/glad/glad_opengl_es_1.0/src/glad.c:78:13: warning: implicit declaration of function ‘free’ [-Wimplicit-function-declaration]
   78 |             free((char *)exts_i[index]);
      |             ^~~~
/home/dbeef/spelunky-psp/vendor/glad/glad_opengl_es_1.0/src/glad.c:78:13: warning: incompatible implicit declaration of built-in function ‘free’
/home/dbeef/spelunky-psp/vendor/glad/glad_opengl_es_1.0/src/glad.c:26:1: note: include ‘<stdlib.h>’ or provide a declaration of ‘free’
   25 | #include <glad/glad.h>
  +++ |+#include <stdlib.h>
   26 | 
/home/dbeef/spelunky-psp/vendor/glad/glad_opengl_es_1.0/src/glad.c:80:9: warning: incompatible implicit declaration of built-in function ‘free’
   80 |         free((void *)exts_i);
      |         ^~~~
/home/dbeef/spelunky-psp/vendor/glad/glad_opengl_es_1.0/src/glad.c:80:9: note: include ‘<stdlib.h>’ or provide a declaration of ‘free’
In file included from /usr/local/pspdev/psp/include/c++/9.3.0/cmath:42,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/math.h:36,
                 from /home/dbeef/spelunky-psp/vendor/stb/include/stb/image.h:516,
                 from /home/dbeef/spelunky-psp/vendor/stb/src/stb_image.cpp:7:
/usr/local/pspdev/psp/include/c++/9.3.0/bits/cpp_type_traits.h:200:12: error: redefinition of ‘struct std::__is_integer<short int>’
  200 |     struct __is_integer<short>
      |            ^~~~~~~~~~~~~~~~~~~
/usr/local/pspdev/psp/include/c++/9.3.0/bits/cpp_type_traits.h:167:12: note: previous definition of ‘struct std::__is_integer<short int>’
  167 |     struct __is_integer<wchar_t>
      |            ^~~~~~~~~~~~~~~~~~~~~
In file included from /usr/local/pspdev/psp/include/c++/9.3.0/bits/std_abs.h:38,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/cmath:47,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/math.h:36,
                 from /home/dbeef/spelunky-psp/vendor/stb/include/stb/image.h:516,
                 from /home/dbeef/spelunky-psp/vendor/stb/src/stb_image.cpp:7:
/usr/local/pspdev/psp/include/stdlib.h:33:3: error: conflicting declaration ‘typedef struct div_t div_t’
   33 | } div_t;
      |   ^~~~~
In file included from /home/dbeef/spelunky-psp/vendor/stb/include/stb/image.h:511,
                 from /home/dbeef/spelunky-psp/vendor/stb/src/stb_image.cpp:7:
/usr/local/pspdev/psp/sdk/include/libc/stdlib.h:47:3: note: previous declaration as ‘typedef struct div_t div_t’
   47 | } div_t;
      |   ^~~~~
In file included from /usr/local/pspdev/psp/include/c++/9.3.0/bits/std_abs.h:38,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/cmath:47,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/math.h:36,
                 from /home/dbeef/spelunky-psp/vendor/stb/include/stb/image.h:516,
                 from /home/dbeef/spelunky-psp/vendor/stb/src/stb_image.cpp:7:
/usr/local/pspdev/psp/include/stdlib.h:39:3: error: conflicting declaration ‘typedef struct ldiv_t ldiv_t’
   39 | } ldiv_t;
      |   ^~~~~~
In file included from /home/dbeef/spelunky-psp/vendor/stb/include/stb/image.h:511,
                 from /home/dbeef/spelunky-psp/vendor/stb/src/stb_image.cpp:7:
/usr/local/pspdev/psp/sdk/include/libc/stdlib.h:57:3: note: previous declaration as ‘typedef struct ldiv_t ldiv_t’
   57 | } ldiv_t;
      |   ^~~~~~
[ 14%] Linking C static library libLogger.a
In file included from /usr/local/pspdev/psp/include/c++/9.3.0/ratio:38,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/chrono:38,
                 from /home/dbeef/spelunky-psp/src/time/interface/time/Timestep.hpp:4,
                 from /home/dbeef/spelunky-psp/src/time/src/Timestep.cpp:2:
/usr/local/pspdev/psp/include/c++/9.3.0/type_traits:252:12: error: redefinition of ‘struct std::__is_integral_helper<short int>’
  252 |     struct __is_integral_helper<short>
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/pspdev/psp/include/c++/9.3.0/type_traits:233:12: note: previous definition of ‘struct std::__is_integral_helper<short int>’
  233 |     struct __is_integral_helper<wchar_t>
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cd /home/dbeef/spelunky-psp/tmp/build-psp/vendor/logger && /usr/bin/cmake -P CMakeFiles/Logger.dir/cmake_clean_target.cmake
/usr/local/pspdev/psp/include/c++/9.3.0/type_traits:1653:12: error: redefinition of ‘struct std::__make_unsigned<short int>’
 1653 |     struct __make_unsigned<wchar_t>
      |            ^~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/pspdev/psp/include/c++/9.3.0/type_traits:1556:12: note: previous definition of ‘struct std::__make_unsigned<short int>’
 1556 |     struct __make_unsigned<short>
      |            ^~~~~~~~~~~~~~~~~~~~~~
cd /home/dbeef/spelunky-psp/tmp/build-psp/vendor/logger && /usr/bin/cmake -E cmake_link_script CMakeFiles/Logger.dir/link.txt --verbose=1
In file included from /usr/local/pspdev/psp/include/c++/9.3.0/chrono:40,
                 from /home/dbeef/spelunky-psp/src/time/interface/time/Timestep.hpp:4,
                 from /home/dbeef/spelunky-psp/src/time/src/Timestep.cpp:2:
/usr/local/pspdev/psp/include/c++/9.3.0/limits:920:12: error: redefinition of ‘struct std::numeric_limits<short int>’
  920 |     struct numeric_limits<short>
      |            ^~~~~~~~~~~~~~~~~~~~~
/usr/local/pspdev/psp/include/c++/9.3.0/limits:663:12: note: previous definition of ‘struct std::numeric_limits<short int>’
  663 |     struct numeric_limits<wchar_t>
      |            ^~~~~~~~~~~~~~~~~~~~~~~
/usr/local/pspdev/bin/psp-ar qc libLogger.a  CMakeFiles/Logger.dir/src/log.c.obj
/usr/local/pspdev/bin/psp-ranlib libLogger.a
make[2]: Leaving directory `/home/dbeef/spelunky-psp/tmp/build-psp'
[ 14%] Built target Logger
/home/dbeef/Android/ndk-standalone-24-arm/bin/make -f vendor/cjson/CMakeFiles/cjson.dir/build.make vendor/cjson/CMakeFiles/cjson.dir/depend
make[2]: Entering directory `/home/dbeef/spelunky-psp/tmp/build-psp'
cd /home/dbeef/spelunky-psp/tmp/build-psp && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/dbeef/spelunky-psp /home/dbeef/spelunky-psp/vendor/cjson /home/dbeef/spelunky-psp/tmp/build-psp /home/dbeef/spelunky-psp/tmp/build-psp/vendor/cjson /home/dbeef/spelunky-psp/tmp/build-psp/vendor/cjson/CMakeFiles/cjson.dir/DependInfo.cmake --color=
In file included from /usr/local/pspdev/psp/include/c++/9.3.0/bits/stl_algobase.h:61,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/memory:62,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/thread:39,
                 from /home/dbeef/spelunky-psp/src/time/interface/time/Timestep.hpp:5,
                 from /home/dbeef/spelunky-psp/src/time/src/Timestep.cpp:2:
/usr/local/pspdev/psp/include/c++/9.3.0/bits/cpp_type_traits.h:200:12: error: redefinition of ‘struct std::__is_integer<short int>’
  200 |     struct __is_integer<short>
      |            ^~~~~~~~~~~~~~~~~~~
/usr/local/pspdev/psp/include/c++/9.3.0/bits/cpp_type_traits.h:167:12: note: previous definition of ‘struct std::__is_integer<short int>’
  167 |     struct __is_integer<wchar_t>
      |            ^~~~~~~~~~~~~~~~~~~~~
Scanning dependencies of target cjson
make[2]: Leaving directory `/home/dbeef/spelunky-psp/tmp/build-psp'
/home/dbeef/Android/ndk-standalone-24-arm/bin/make -f vendor/cjson/CMakeFiles/cjson.dir/build.make vendor/cjson/CMakeFiles/cjson.dir/build
make[2]: Entering directory `/home/dbeef/spelunky-psp/tmp/build-psp'
[ 17%] Building C object vendor/cjson/CMakeFiles/cjson.dir/cJSON.c.obj
cd /home/dbeef/spelunky-psp/tmp/build-psp/vendor/cjson && /usr/local/pspdev/bin/psp-gcc -DCJSON_API_VISIBILITY -DCJSON_EXPORT_SYMBOLS -DENABLE_LOCALES -DHAVE_OPENGL -D__PSP__ -Dwchar_t=short -I/home/dbeef/spelunky-psp/.. -I/usr/local/pspdev/psp/sdk/include -I/home/dbeef/spelunky-psp/vendor/cjson  -D_PSP_FW_VERSION=999999 -Wextra -Werror -Wstrict-prototypes -Wwrite-strings -Winit-self -Wcast-align -Wformat=2 -Wmissing-prototypes -Wstrict-overflow=2 -Wcast-qual -Wundef -Wswitch-default -Wconversion -Wc++-compat -Wdouble-promotion -Wparentheses -Wformat-overflow -Wunused-macros -Wswitch-enum -fvisibility=hidden  -O3 -DNDEBUG   -G0 -I/usr/local/pspdev/psp/include -I/usr/local/pspdev/psp/sdk/include/libc -G0 -I/usr/local/pspdev/psp/include -I/usr/local/pspdev/psp/sdk/include/libc -std=gnu11 -o CMakeFiles/cjson.dir/cJSON.c.obj   -c /home/dbeef/spelunky-psp/vendor/cjson/cJSON.c
[ 20%] Linking C static library libglad.a
cd /home/dbeef/spelunky-psp/tmp/build-psp/vendor/glad/glad_opengl_es_1.0 && /usr/bin/cmake -P CMakeFiles/glad.dir/cmake_clean_target.cmake
cd /home/dbeef/spelunky-psp/tmp/build-psp/vendor/glad/glad_opengl_es_1.0 && /usr/bin/cmake -E cmake_link_script CMakeFiles/glad.dir/link.txt --verbose=1
In file included from /home/dbeef/spelunky-psp/vendor/cjson/cJSON.c:41:
/usr/local/pspdev/psp/sdk/include/libc/stdio.h: In function ‘lseek’:
/usr/local/pspdev/psp/sdk/include/libc/stdio.h:36:79: error: conversion from ‘SceOff’ {aka ‘long long int’} to ‘off_t’ {aka ‘long int’} may change value [-Werror=conversion]
   36 | static __inline__ off_t lseek(int handle, off_t position, int wheel) { return sceIoLseek(handle, position, wheel); }
      |                                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/pspdev/psp/sdk/include/libc/stdio.h: In function ‘tell’:
/usr/local/pspdev/psp/sdk/include/libc/stdio.h:37:51: error: conversion from ‘SceOff’ {aka ‘long long int’} to ‘off_t’ {aka ‘long int’} may change value [-Werror=conversion]
   37 | static __inline__ off_t tell(int handle) { return sceIoLseek(handle, 0, 1); }
      |                                                   ^~~~~~~~~~~~~~~~~~~~~~~~
/home/dbeef/spelunky-psp/vendor/cjson/cJSON.c: At top level:
/home/dbeef/spelunky-psp/vendor/cjson/cJSON.c:149:25: error: ‘malloc’ undeclared here (not in a function)
  149 | #define internal_malloc malloc
      |                         ^~~~~~
/home/dbeef/spelunky-psp/vendor/cjson/cJSON.c:157:40: note: in expansion of macro ‘internal_malloc’
  157 | static internal_hooks global_hooks = { internal_malloc, internal_free, internal_realloc };
      |                                        ^~~~~~~~~~~~~~~
/home/dbeef/spelunky-psp/vendor/cjson/cJSON.c:59:1: note: ‘malloc’ is defined in header ‘<stdlib.h>’; did you forget to ‘#include <stdlib.h>’?
   58 | #include "cJSON.h"
  +++ |+#include <stdlib.h>
   59 | 
/home/dbeef/spelunky-psp/vendor/cjson/cJSON.c:150:23: error: ‘free’ undeclared here (not in a function)
  150 | #define internal_free free
      |                       ^~~~
/home/dbeef/spelunky-psp/vendor/cjson/cJSON.c:157:57: note: in expansion of macro ‘internal_free’
  157 | static internal_hooks global_hooks = { internal_malloc, internal_free, internal_realloc };
      |                                                         ^~~~~~~~~~~~~
/home/dbeef/spelunky-psp/vendor/cjson/cJSON.c:150:23: note: ‘free’ is defined in header ‘<stdlib.h>’; did you forget to ‘#include <stdlib.h>’?
  150 | #define internal_free free
      |                       ^~~~
/home/dbeef/spelunky-psp/vendor/cjson/cJSON.c:157:57: note: in expansion of macro ‘internal_free’
  157 | static internal_hooks global_hooks = { internal_malloc, internal_free, internal_realloc };
      |                                                         ^~~~~~~~~~~~~
/home/dbeef/spelunky-psp/vendor/cjson/cJSON.c:151:26: error: ‘realloc’ undeclared here (not in a function)
  151 | #define internal_realloc realloc
      |                          ^~~~~~~
/home/dbeef/spelunky-psp/vendor/cjson/cJSON.c:157:72: note: in expansion of macro ‘internal_realloc’
  157 | static internal_hooks global_hooks = { internal_malloc, internal_free, internal_realloc };
      |                                                                        ^~~~~~~~~~~~~~~~
/home/dbeef/spelunky-psp/vendor/cjson/cJSON.c:151:26: note: ‘realloc’ is defined in header ‘<stdlib.h>’; did you forget to ‘#include <stdlib.h>’?
  151 | #define internal_realloc realloc
      |                          ^~~~~~~
/home/dbeef/spelunky-psp/vendor/cjson/cJSON.c:157:72: note: in expansion of macro ‘internal_realloc’
  157 | static internal_hooks global_hooks = { internal_malloc, internal_free, internal_realloc };
      |                                                                        ^~~~~~~~~~~~~~~~
/usr/local/pspdev/bin/psp-ar qc libglad.a  CMakeFiles/glad.dir/src/glad.c.obj
In file included from /usr/local/pspdev/psp/include/c++/9.3.0/bits/localefwd.h:42,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/string:43,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/stdexcept:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/array:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/tuple:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/bits/unique_ptr.h:37,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/memory:80,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/thread:39,
                 from /home/dbeef/spelunky-psp/src/time/interface/time/Timestep.hpp:5,
                 from /home/dbeef/spelunky-psp/src/time/src/Timestep.cpp:2:
/usr/local/pspdev/psp/include/c++/9.3.0/cctype:87:11: error: ‘::isblank’ has not been declared
   87 |   using ::isblank;
      |           ^~~~~~~
/usr/local/pspdev/bin/psp-ranlib libglad.a
make[2]: Leaving directory `/home/dbeef/spelunky-psp/tmp/build-psp'
cc1: all warnings being treated as errors
make[2]: *** [vendor/cjson/CMakeFiles/cjson.dir/cJSON.c.obj] Error 1
make[2]: Leaving directory `/home/dbeef/spelunky-psp/tmp/build-psp'
make[1]: *** [vendor/cjson/CMakeFiles/cjson.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
In file included from /usr/local/pspdev/psp/include/c++/9.3.0/cstdlib:75,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/ext/string_conversions.h:41,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/bits/basic_string.h:6493,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/string:55,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/stdexcept:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/array:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/tuple:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/bits/unique_ptr.h:37,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/memory:80,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/thread:39,
                 from /home/dbeef/spelunky-psp/src/time/interface/time/Timestep.hpp:5,
                 from /home/dbeef/spelunky-psp/src/time/src/Timestep.cpp:2:
/usr/local/pspdev/psp/include/stdlib.h:33:3: error: conflicting declaration ‘typedef struct div_t div_t’
   33 | } div_t;
      |   ^~~~~
In file included from /usr/local/pspdev/psp/include/SDL/SDL_stdinc.h:40,
                 from /usr/local/pspdev/psp/include/SDL/SDL_timer.h:30,
                 from /home/dbeef/spelunky-psp/src/time/src/Timestep.cpp:1:
/usr/local/pspdev/psp/sdk/include/libc/stdlib.h:47:3: note: previous declaration as ‘typedef struct div_t div_t’
   47 | } div_t;
      |   ^~~~~
In file included from /usr/local/pspdev/psp/include/c++/9.3.0/cstdlib:75,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/ext/string_conversions.h:41,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/bits/basic_string.h:6493,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/string:55,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/stdexcept:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/array:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/tuple:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/bits/unique_ptr.h:37,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/memory:80,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/thread:39,
                 from /home/dbeef/spelunky-psp/src/time/interface/time/Timestep.hpp:5,
                 from /home/dbeef/spelunky-psp/src/time/src/Timestep.cpp:2:
/usr/local/pspdev/psp/include/stdlib.h:39:3: error: conflicting declaration ‘typedef struct ldiv_t ldiv_t’
   39 | } ldiv_t;
      |   ^~~~~~
In file included from /usr/local/pspdev/psp/include/SDL/SDL_stdinc.h:40,
                 from /usr/local/pspdev/psp/include/SDL/SDL_timer.h:30,
                 from /home/dbeef/spelunky-psp/src/time/src/Timestep.cpp:1:
/usr/local/pspdev/psp/sdk/include/libc/stdlib.h:57:3: note: previous declaration as ‘typedef struct ldiv_t ldiv_t’
   57 | } ldiv_t;
      |   ^~~~~~
[ 20%] Built target glad
In file included from /usr/local/pspdev/psp/include/c++/9.3.0/bits/basic_string.h:6719,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/string:55,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/stdexcept:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/array:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/tuple:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/bits/unique_ptr.h:37,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/memory:80,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/thread:39,
                 from /home/dbeef/spelunky-psp/src/time/interface/time/Timestep.hpp:5,
                 from /home/dbeef/spelunky-psp/src/time/src/Timestep.cpp:2:
/usr/local/pspdev/psp/include/c++/9.3.0/bits/functional_hash.h:150:3: error: redefinition of ‘struct std::hash<short int>’
  150 |   _Cxx_hashtable_define_trivial_hash(short)
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/pspdev/psp/include/c++/9.3.0/bits/functional_hash.h:136:3: note: previous definition of ‘struct std::hash<short int>’
  136 |   _Cxx_hashtable_define_trivial_hash(wchar_t)
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/local/pspdev/psp/include/c++/9.3.0/string:55,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/stdexcept:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/array:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/tuple:39,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/bits/unique_ptr.h:37,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/memory:80,
                 from /usr/local/pspdev/psp/include/c++/9.3.0/thread:39,
                 from /home/dbeef/spelunky-psp/src/time/interface/time/Timestep.hpp:5,
                 from /home/dbeef/spelunky-psp/src/time/src/Timestep.cpp:2:
/usr/local/pspdev/psp/include/c++/9.3.0/bits/basic_string.h:6825:5: error: ‘std::__cxx11::basic_string<short int> std::literals::string_literals::operator""s(const short int*, std::size_t)’ has invalid argument list
 6825 |     operator""s(const wchar_t* __str, size_t __len)
      |     ^~~~~~~~
make[2]: *** [vendor/stb/CMakeFiles/stb.dir/src/stb_image.cpp.obj] Error 1
make[2]: Leaving directory `/home/dbeef/spelunky-psp/tmp/build-psp'
make[1]: *** [vendor/stb/CMakeFiles/stb.dir/all] Error 2
make[2]: *** [src/time/CMakeFiles/Time.dir/src/Timestep.cpp.obj] Error 1
make[2]: Leaving directory `/home/dbeef/spelunky-psp/tmp/build-psp'
make[1]: *** [src/time/CMakeFiles/Time.dir/all] Error 2
```

**Comment by dbeef** [comment]
All build problems come from `wchar_t` being alias for short, thus making overloaded functions being actually redefinitions causing build error.

**Comment by dbeef** [comment]
Ok I am calling that off, it was problem with my cmake toolchain file for PSP that manifested because they supposedly added `wchar_t` support in CPP's `<limits>`:

```
./cmake/toolchains/PSP.cmake:37:ADD_DEFINITIONS("-Dwchar_t=short")
```
Removing that line fixes the problem.

**Comment by dbeef** [comment]
Spelunky PSP built with new toolchain totally work ;)

image

File: scripts/002-gcc-stage1.sh [hunk 1]
```diff
@@ -26,13 +28,15 @@
...
-
+ ln -fs ../isl-$ISL_VERSION isl
+ 
  ## Create and enter the build directory.
  mkdir build-psp
  cd build-psp
 
  ## Configure the build.
- CFLAGS="$CFLAGS -I/opt/local/include" CPPFLAGS="$CPPFLAGS -I/opt/local/include" LDFLAGS="$LDFLAGS -L/opt/local/lib" ../configure --prefix="$PSPDEV" --target="psp" --enable-languages="c" --enable-lto --with-newlib --with-gmp-include="$(pwd)/gmp" --with-gmp-lib="$(pwd)/gmp/.libs" --with-mpfr-include="$(pwd)/../mpfr/src" --with-mpfr-lib="$(pwd)/mpfr/src/.libs" --without-headers --disable-libssp
+ #CFLAGS="$CFLAGS -I/opt/local/include" CPPFLAGS="$CPPFLAGS -I/opt/local/include" LDFLAGS="$LDFLAGS -L/opt/local/lib" ../configure --prefix="$PSPDEV" --target="psp" --enable-languages="c" --enable-lto --with-newlib --with-gmp-include="$(pwd)/gmp" --with-gmp-lib="$(pwd)/gmp/.libs" --with-mpfr-include="$(pwd)/../mpfr/src" --with-mpfr-lib="$(pwd)/mpfr/src/.libs" --without-headers --disable-libssp
```
Comment (dbeef) [inline scripts/002-gcc-stage1.sh:38]: Remove this line before merging

File: scripts/005-gcc-stage2.sh [hunk 2]
```diff
@@ -2,40 +2,44 @@
 # gcc-stage2.sh by Naomi Peori (naomi@peori.ca) customized by yreeen (yreeen@gmail.com)
 # gdc support from TurkeyMan( https://github.com/TurkeyMan )
 
- ## set gcc version
- GCC_VERSION=4.9.3
- GMP_VERSION=5.1.3
- MPC_VERSION=1.0.2
- MPFR_VERSION=3.1.2
+## set gcc version
```
Comment (dbeef) [inline scripts/005-gcc-stage2.sh:5]: Keep indentation consistent

File: scripts/005-gcc-stage2.sh [hunk 3]
```diff
@@ -2,40 +2,44 @@
...
  download_and_extract https://ftp.gnu.org/gnu/gmp/gmp-$GMP_VERSION.tar.bz2 gmp-$GMP_VERSION
  download_and_extract https://ftp.gnu.org/gnu/mpc/mpc-$MPC_VERSION.tar.gz mpc-$MPC_VERSION
  download_and_extract https://ftp.gnu.org/gnu/mpfr/mpfr-$MPFR_VERSION.tar.bz2 mpfr-$MPFR_VERSION
+ download_and_extract http://isl.gforge.inria.fr/isl-$ISL_VERSION.tar.gz isl-$ISL_VERSION
 
  ## Enter the source directory and patch the source code.
  cd gcc-$GCC_VERSION
  patch -p1 -i ../../patches/gcc-$GCC_VERSION-PSP.patch
- patch -p0 -i ../../patches/patch-gcc_cp_cfns.h
+# patch -p0 -i ../../patches/patch-gcc_cp_cfns.h
```
Comment (dbeef) [inline scripts/005-gcc-stage2.sh:27]: Remove this line before merging

File: scripts/005-gcc-stage2.sh [hunk 4]
```diff
@@ -2,40 +2,44 @@
...
+ ln -fs ../isl-$ISL_VERSION isl
 
  ## Create and enter the build directory.
  mkdir build-psp
  cd build-psp
 
  ## Configure the build.
  ## If you want to build gdc add "d" to --enable-languages option.
- CFLAGS="$CFLAGS -I/opt/local/include" CPPFLAGS="$CPPFLAGS -I/opt/local/include" LDFLAGS="$LDFLAGS -L/opt/local/lib" ../configure --prefix="$PSPDEV" --target="psp" --enable-languages="c,c++" --enable-lto --with-newlib --with-gmp-include="$(pwd)/gmp" --with-gmp-lib="$(pwd)/gmp/.libs" --with-mpfr-include="$(pwd)/../mpfr/src" --with-mpfr-lib="$(pwd)/mpfr/src/.libs" --enable-cxx-flags="-G0"
+ #CFLAGS="$CFLAGS -I/opt/local/include" CPPFLAGS="$CPPFLAGS -I/opt/local/include" LDFLAGS="$LDFLAGS -L/opt/local/lib" ../configure --prefix="$PSPDEV" --target="psp" --enable-languages="c,c++" --enable-lto --with-newlib --with-gmp-include="$(pwd)/gmp" --with-gmp-lib="$(pwd)/gmp/.libs" --with-mpfr-include="$(pwd)/../mpfr/src" --with-mpfr-lib="$(pwd)/mpfr/src/.libs" --enable-cxx-flags="-G0"
```
Comment (dbeef) [inline scripts/005-gcc-stage2.sh:41]: Remove line

**Comment by xantares** [comment]
hi @dbeef could you update to gcc 10.1.0 ?

**Comment by sharkwouter** [comment]
@xantares is there something 10.1.0 provides which you require?

**Comment by xantares** [comment]
I was hoping to test c++20

**Comment by sharkwouter** [comment]
Ah okay, I think the version currently in the toolchain also support C++ 20, but it might not be stable. I don't think we'll update to the new version soon, though, it's a lot of work to make sure it works well.

**Pull request merged**